File: 03_liskov_substitution/typescript/good/index01.good.ts
class Shape {
  constructor(width: number, height: number) {
    this.width = width
    this.height = height
  }
  area () { }
  setWidth (width: number) { }
  setHeight (height: number) { }
  width: number
  height: number
}

class Rectangle extends Shape {
  constructor(width: number, height: number) {
    super(width,height)
  }

  setWidth (width: number) {
    this.width = width
  }
  setHeight (height: number) {
    this.height = height
  }

  area (): number {
    return this.width * this.height
  }
}
class Square extends Shape {
  setWidth (width: number) {
    this.width = width
    this.height = width
  }
  setHeight (height: number) {
    this.width = height
    this.height = height
  }
  area (): number {
    return this.width * this.height
  }
}

function increaseReactangleWidth (rectangle: Rectangle) {
  rectangle.setWidth(rectangle.width + 1)
}
function increaseSquareWidth(square: Square) {
  square.setWidth(square.width + 1);
}

const rectangle1 = new Rectangle(10, 2)
const square = new Square(5, 5)

increaseReactangleWidth(rectangle1)
increaseSquareWidth(square)

console.log(rectangle1.area())
console.log(square.area())

File: 03_liskov_substitution/dart/good/index01.good.dart
abstract class Shape {
  Shape(this.width, this.height);
  late int width;
  late int height;
  void setWidth(int width);
  void setHeight(int height);
  int area();
}

class Rectangle extends Shape {
  Rectangle(super.width, super.height);

  @override
  void setWidth(width) => this.width = width;

  @override
  void setHeight(height) => this.height = height;

  int area() => this.width * this.height;
}

class Square extends Shape {
  Square(super.width, super.height);

  setWidth(int width) {
    this.width = width;
    this.height = width;
  }

  setHeight(int height) {
    this.width = height;
    this.height = height;
  }

  int area() => this.width * this.height;
}

increaseReactangleWidth(Rectangle rectangle) {
  rectangle.setWidth(rectangle.width + 1);
}

increaseSquareWidth(Square square) {
  square.setWidth(square.width + 1);
}

void main(List<String> args) {
  var rectangle1 = Rectangle(10, 2);
  var square = Square(5, 5);

  increaseReactangleWidth(rectangle1);
  increaseSquareWidth(square);

  print(rectangle1.area());
  print(square.area());
}
